### PR TITLE
Added click-boxes of `show` and `locked` 

### PR DESCRIPTION
**Pull request by PingHsunTsai**
#146 
### What type of change is this?

- [ ] * Added click-boxes of `show` and `locked` for `compas_viewer.components.SceneForm`.
- [ ] * Added observer pattern to of `show` and `locked` for `compas_viewer.Scene.scene`.
- [ ] * Added Time Debounce pattern to of `show` and `locked` for `compas_viewer.Scene.scene`.

Screenshot from 2024-05-27 17-48-08

### CODE REVIEW

**Comment by Licini** [comment]
This is nice, only thing is that, I feel `is_locked` is not necessary to be shown here, what you think? @tomvanmele

**Comment by Licini** [comment]
I don't know even is_locked is useful at all in any cases, as this is not aiming on being a CAD environment that needs to do all kinds editing

**Comment by tomvanmele** [comment]
> I don't know even is_locked is useful at all in any cases, as this is not aiming on being a CAD environment that needs to do all kinds editing

yes, let's get rid of `is_locked` completely...

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Added `clear_scene` action.
 * Added `compas_viewer.components.SceneForm`.
 * Added `callback` for `compas_viewer.components.TreeForm`.
+* Added click-boxes of `show` and `locked` for `compas_viewer.components.SceneForm`.
+* Added observer pattern to of `show` and `locked` for `compas_viewer.Scene.scene`.
+* Added Time Debounce pattern to of `show` and `locked` for `compas_viewer.Scene.scene`.
```
Comment (Licini) [inline CHANGELOG.md:21]: remove locked here

File: src/compas_viewer/components/sceneform.py [hunk 2]
```diff
@@ -101,42 +111,61 @@ def scene(self, scene: Scene):
...
-    def on_item_clickded(self):
-        selected_nodes = [item.node for item in self.selectedItems()]
-        for node in self.scene.objects:
-            node.is_selected = node in selected_nodes
-            if self.callback and node.is_selected:
-                self.callback(node)
+    def on_item_clicked(self, item, column):
+        if column == self.show_idx:
+            is_visible = item.checkState(self.show_idx) == Qt.Checked
+            item.node.show = is_visible
```
Comment (Licini) [inline src/compas_viewer/components/sceneform.py:152]: can be one line

File: src/compas_viewer/components/sceneform.py [hunk 3]
```diff
@@ -101,42 +111,61 @@ def scene(self, scene: Scene):
...
+        self.viewer.renderer.update()
 
     def on_item_selection_changed(self):
         for item in self.selectedItems():
             if self.callback:
                 self.callback(item.node)
+
+    def adjust_column_widths(self, item=None, column=None):
+        for i in range(self.columnCount()):
+            self.resizeColumnToContents(i)
```
Comment (Licini) [inline src/compas_viewer/components/sceneform.py:171]: Would it be possible to shrink the column towards right side? like this:
image

File: src/compas_viewer/scene/bufferobject.py [hunk 4]
```diff
@@ -208,12 +209,27 @@ def __init__(
...
+    @is_locked.setter
+    def is_locked(self, value: bool):
+        self._is_locked = value
+        if value:
+            self.is_selected = False
+            # scene parent
+            self.scene.instance_colors.pop(self.instance_color.rgb255)
+        else:
+            self.scene.instance_colors[self.instance_color.rgb255] = self
+
```
Comment (Licini) [inline src/compas_viewer/scene/bufferobject.py:232]: We will be removing is_locked all together, but will be another PR.

File: src/compas_viewer/scene/scene.py [hunk 5]
```diff
@@ -70,10 +72,29 @@ def __init__(self, name: str = "ViewerScene", context: str = "Viewer"):
...
+        self.update_timer = QTimer()
+        self.update_timer.setSingleShot(True)
+        self.update_timer.timeout.connect(self.update_observers)
+        self.debounce_interval = 200
+
+    @property
+    def observers(self):
+        from compas_viewer import Viewer
+
+        self.viewer = Viewer()
```
Comment (Licini) [inline src/compas_viewer/scene/scene.py:90]: make this a viewer property?

File: src/compas_viewer/scene/scene.py [hunk 6]
```diff
@@ -70,10 +72,29 @@ def __init__(self, name: str = "ViewerScene", context: str = "Viewer"):
...
+        self._observers.extend(
+            [
+                self.viewer.renderer,
+                self.viewer.ui.sidebar,
+            ]
+        )
```
Comment (Licini) [inline src/compas_viewer/scene/scene.py:96]: I think this will repetitively expand the `_observer` list. You need to either make it a `Set` or check if things has already been added to list

File: src/compas_viewer/scene/scene.py [hunk 7]
```diff
@@ -70,10 +72,29 @@ def __init__(self, name: str = "ViewerScene", context: str = "Viewer"):
 
         #  Primitive
         self.objects: list[ViewerSceneObject]
-
+        self._observers = []
```
Comment (Licini) [inline src/compas_viewer/scene/scene.py:75]: Add to Attributes section of docstring what observers is

File: src/compas_viewer/ui/sidebar.py [hunk 8]
```diff
@@ -1,18 +1,32 @@
...
-    def __init__(self, ui: "UI", show: bool = True) -> None:
+    def __init__(self, ui: "UI", show: bool, items: list[dict]) -> None:
         self.ui = ui
-        self.widget = QtWidgets.QSplitter(QtCore.Qt.Orientation.Vertical)
+        self.widget = QSplitter(QtCore.Qt.Orientation.Vertical)
         self.widget.setChildrenCollapsible(True)
         self.widget.setVisible(show)
+        self.add_sidebar(items, self.widget)
+
+    def add_sidebar(self, items: list[dict], parent: QSplitter) -> None:
```
Comment (Licini) [inline src/compas_viewer/ui/sidebar.py:20]: should this be named `add_items` instead?

File: src/compas_viewer/ui/sidebar.py [hunk 9]
```diff
@@ -1,18 +1,32 @@
...
+
+from compas_viewer.components import Sceneform
 
 if TYPE_CHECKING:
     from .ui import UI
 
 
 class SideBarRight:
-    def __init__(self, ui: "UI", show: bool = True) -> None:
+    def __init__(self, ui: "UI", show: bool, items: list[dict]) -> None:
```
Comment (Licini) [inline src/compas_viewer/ui/sidebar.py:13]: `list[dict[str, str]]` ?

File: src/compas_viewer/scene/bufferobject.py (hunk 4 above)
Comment (PingHsunTsai) [reply src/compas_viewer/scene/bufferobject.py:232]: okay!

File: src/compas_viewer/config.py [hunk 10]
```diff
@@ -238,7 +240,17 @@ class StatusbarConfig(ConfigBase):
...
     show: bool = True
     sceneform: bool = True
-    items: list[dict[str, str]] = None
+    items: list[dict[str, str]] = field(
+        default_factory=lambda: [
+            {
+                "type": "Sceneform",
+                "action": {
+                    "Name": get_name,
+                    "Show": get_show_box,
```
Comment (Licini) [inline src/compas_viewer/config.py:249]:
```python
"Name": lambda obj: obj.name,
"Show": {
    "type": "checkbox"
    "getter": lambda obj: obj.show,
    "setter": lambda obj, value: obj.show = value  
}
```

File: src/compas_viewer/commands.py [hunk 11]
```diff
@@ -460,3 +461,23 @@ def load_data():
...
+def get_name(obj: "ViewerSceneObject"):
+    return str(obj.name)
+
+
+def get_show_box(obj: "ViewerSceneObject"):
+    """
+    'checkbox' to show(hide) the ViewerSceneObject.
+    'show' is a boolean attribute of ViewerSceneObject.
+    """
+    return "checkbox", "show"
```
Comment (Licini) [inline src/compas_viewer/commands.py:483]: I think these functions should not be here. I don't feel these two is going to be called else where other than configuring the sceneform.

File: src/compas_viewer/config.py (hunk 10 above)
Comment (PingHsunTsai) [reply src/compas_viewer/config.py:249]:
```
{
    "type": "Sceneform",
    "action": [
        {"title": "Name", "action": get_obj_attr_cmd, "kwargs": {"attr": "name"}},
        {"title": "Show", "type": "checkbox", "action": get_obj_attr_cmd, "kwargs": {"attr": "show"}},
    ],
}
```
I found another solution that has similar structure than other :)

**Review by Licini — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 12]
```diff
@@ -22,6 +22,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Added default colors to `GeometryObject`.
 * Added `object_info_cmd` for `compas_viewer.commends`.
 * Added `gridmode` to `GridObject`.
+* Added click-boxes of `show` for `compas_viewer.components.SceneForm`.
+* Added observer pattern to of `show` for `compas_viewer.Scene.scene`.
+* Added Time Debounce pattern to of `show` for `compas_viewer.Scene.scene`.
```
Comment (tomvanmele) [inline CHANGELOG.md:27]: please convert this into understandable sentences...

File: src/compas_viewer/scene/bufferobject.py [hunk 13]
```diff
@@ -206,7 +207,8 @@ def __init__(
         self.linewidth = 1.0 if linewidth is None else linewidth
         self.opacity = 1.0 if opacity is None else opacity
         self.doublesided = True if doublesided is None else doublesided
-        self.is_visible = True if is_visiable is None else is_visiable
+        self.show = True if is_visiable is None else is_visiable
```
Comment (tomvanmele) [inline src/compas_viewer/scene/bufferobject.py:210]: `is_visiable` => `is_visible`

Comment (tomvanmele) [reply src/compas_viewer/scene/bufferobject.py:210]: btw, i am not a fan of `self.show`. i think it should be `self.is_visible`. `self.show` sounds like a method...

File: src/compas_viewer/scene/bufferobject.py [hunk 14]
```diff
@@ -206,7 +207,8 @@ def __init__(
...
+        self._is_locked = False if is_locked is None else is_locked
```
Comment (tomvanmele) [inline src/compas_viewer/scene/bufferobject.py:211]: were we not going to remove this?

File: src/compas_viewer/scene/scene.py [hunk 15]
```diff
@@ -70,10 +81,35 @@ def __init__(self, name: str = "ViewerScene", context: str = "Viewer"):
 
         #  Primitive
         self.objects: list[ViewerSceneObject]
```
Comment (tomvanmele) [inline src/compas_viewer/scene/scene.py:83]: btw, this does not initialize the attribute. it just adds a type annotation. if you try to access this attribute an error will be thrown. it should be `self.objects: list[ViewerSceneObject] = []`

File: src/compas_viewer/scene/scene.py [hunk 16]
```diff
@@ -70,10 +81,35 @@ def __init__(self, name: str = "ViewerScene", context: str = "Viewer"):
...
+    def observers(self):
+        new_observers = [
+            self.viewer.renderer,
+            self.viewer.ui.sidebar,
+        ]
+
+        for observer in new_observers:
+            self._observers.add(observer)
+
+        return list(self._observers)
```
Comment (tomvanmele) [inline src/compas_viewer/scene/scene.py:112]: this looks a bit fishy...

File: src/compas_viewer/scene/scene.py [hunk 17]
```diff
@@ -60,6 +62,15 @@ class ViewerScene(Scene):
     context : str, optional
         The context of the scene.
 
+    Attributes
+    ----------
+    observers : set
+        Set of unique observer objects.
```
Comment (tomvanmele) [inline src/compas_viewer/scene/scene.py:68]: are these objects that observe, or objects that are being observed?

what makes something an "observer"?

is there a base class somewhere?

File: src/compas_viewer/scene/bufferobject.py (hunk 13 above)
Comment (Licini) [reply src/compas_viewer/scene/bufferobject.py:210]: `show` comes from compas core, which I suggested a while ago for the simplicity, but I agree indeed it can feel like a method. If we decide to go back to `is_visible`, I would suggest we make this change in compas core first then update here

File: src/compas_viewer/scene/bufferobject.py (hunk 14 above)
Comment (Licini) [reply src/compas_viewer/scene/bufferobject.py:211]: yes we are, but it would be better to do that in another PR, cuz this attribute is a bit everywhere, this one here is added temporarily to make the form work

File: src/compas_viewer/scene/scene.py (hunk 15 above)
Comment (Licini) [reply src/compas_viewer/scene/scene.py:83]: `objects` is inherent from the parent class, I believe this was only for additionally type-hinting the object will be `ViewerScenceObject` instead of the base `SceneObject`

File: src/compas_viewer/scene/scene.py (hunk 17 above)
Comment (PingHsunTsai) [reply src/compas_viewer/scene/scene.py:68]: I updated the observer class in  PR #181 maybe I remove from here, so is less confusing
